colcon_core/shell/template/__init__.py

```diff
@@ -69,25 +70,31 @@ def installProxy(self):  # noqa: D102 N802
 
 
 class CachingInterpreter(BypassStdoutInterpreter):
-    """Interpreter for EmPy which which caches parsed tokens."""
-
-    def parse(self, scanner, locals=None):  # noqa: A002 D102
-        global cached_tokens
-        data = scanner.buffer
-        # try to use cached tokens
-        tokens = cached_tokens.get(data)
-        if tokens is None:
-            # collect tokens and cache them
-            tokens = []
-            while True:
-                token = scanner.one()
-                if token is None:
-                    break
-                tokens.append(token)
-            cached_tokens[data] = tokens
-
-        # reimplement the parse method using the (cached) tokens
-        self.invoke('atParse', scanner=scanner, locals=locals)
-        for token in tokens:
-            self.invoke('atToken', token=token)
-            token.run(self, locals)
+    """Interpreter for EmPy which caches parsed tokens."""
+
+    class _CachingScannerDecorator(GenericDecorator):
+
+        def __init__(self, decoree, cache):
+            super().__init__(decoree, _cache=cache, _idx=0)
+
+        def one(self, *args, **kwargs):
+            try:
+                token, count = self._cache[self._idx]
+            except IndexError:
+                count = len(self._decoree)
+                token = self._decoree.one(*args, **kwargs)
+                count -= len(self._decoree)
+                self._cache.append((token, count))
+            else:
+                self.advance(count)
+                self.sync()
```

Comment:
This is kind of a preference, but I don't love doing a bunch of work inside of an exception handler that can throw its own exceptions. I know Python 3 handles nested exceptions relatively well now, but I still always find it easier to debug top-level exceptions rather than nested ones. Thus I would rewrite this as:

```suggestion
            if self._idx < len(self._cache):
                token, count = self._cache[self._idx]
                self.advance(count)
                self.sync()
            else:
                count = len(self._decoree)
                token = self._decoree.one(*args, **kwargs)
                count -= len(self._decoree)
                self._cache.append((token, count))
```

(this also lets us group together all of the functionality we need when the index exists in the cache).  But take this as a suggestion; if you prefer using the IndexError, that's fine.